### PR TITLE
React react1 week3

### PR DESCRIPTION
**Pull request by faheem102**

### CODE REVIEW

File: react/react1/week3/time-registration/src/SubmitShifts.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+          type="datetime-local"
+          value={endInput}
+          onChange={(e) => setEndInput(e.target.value)}
+        ></input>
+      </form>
+      <button className="button" onClick={addShift}>
+        Save Shift
+      </button>
+    </div>
+  );
```
Comment (BjornFridal) [inline react/react1/week3/time-registration/src/SubmitShifts.js:75]: Assigning FormPart to a variable inside a component works, but one of the main benefits of using React is that we can create individual components. Doing so gives us more control over our code and it makes it easier to resuse our components.

You can convert FormParts into it's own component like this:

```
const FormPart = ({ setList }) => {
  const [nameInput, setNameInput] = useState('');
  const [startInput, setStartInput] = useState('');
  const [endInput, setEndInput] = useState('');

  const addShift = () => {
    const newId = new Date().getTime();
    const newTodo = {
      name: nameInput,
      start: startInput,
      end: endInput,
      id: newId
    };
    setList((prevTodos) => {
      return [...prevTodos, newTodo];
    });
  };

  return (
    <div>
      <h2>Submit Shift</h2>
      <form>
        <label>Name</label>
        <input
          type="text"
          placeholder="Name"
          value={nameInput}
          onChange={(e) => setNameInput(e.target.value)}
        ></input>
        <label>Start Time</label>
        <input
          type="datetime-local"
          value={startInput}
          onChange={(e) => setStartInput(e.target.value)}
        ></input>
        <label>End Time</label>
        <input
          type="datetime-local"
          value={endInput}
          onChange={(e) => setEndInput(e.target.value)}
        ></input>
      </form>
      <button className="button" onClick={addShift}>
        Save Shift
      </button>
    </div>
  );
};
```

As you can see this also allows us to place all of the useState inside of that component. We can keep the state relevant to our component inside that component. The only piece of code we need is `setList` and we can pass that in using props:

```
<FancyBorder>
  <FormPart setList={setList} />
</FancyBorder>

```

You can do the same with `ShiftsPart` and `display`.

File: react/react1/week3/time-registration/src/SubmitShifts.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      key={shift.newId}
+    />
+  ));
+  const searchEmployee = (e) => {
+    const value = e.target.value.toLowerCase();
+    const searchName = list.filter((shift) =>
+      shift.name.toLowerCase().includes(value)
+    );
+    setList(searchName);
+  };
```
Comment (BjornFridal) [inline react/react1/week3/time-registration/src/SubmitShifts.js:45]: You search works nicely. The only issue is that once you have searched once, you can't undo that search.

Maybe you can keep a copy of the orignal data you fetched and use that for searching?

Here is an example where I keep a copy of the original data I have fetched. Let me know if it makes sense :)

https://codesandbox.io/s/amazing-diffie-crf0wp?file=/src/App.js

File: react/react1/week3/my-app/src/TodoList.js [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+          <input
+            type="date"
+            id="deadline"
+            value={inputDeadline}
+            onChange={(e) => setInputDeadline(e.target.value)}
+          ></input>
+        </form>
+      </div>
+    </FormBorder>
+  );
```
Comment (BjornFridal) [inline react/react1/week3/my-app/src/TodoList.js:160]: I would encourage you to also use individual components here, like I mentioned in the feedback for the **time-registration** application :)

**Comment by BjornFridal** [comment]
Good work Faheem! 👌 The most important feedback from me would be to split more of you code into components. It will make you code much more readable, easier to maintain and reuse. Let me know if you have any questions and I will glady help out 🙂

File: react/react1/week3/time-registration/src/SubmitShifts.js (hunk 1 above)
Comment (faheem102) [reply react/react1/week3/time-registration/src/SubmitShifts.js:75]: hi Bjorn thank you for your feedback. the method that you have suggested is very efficient. but I have confusion about which props should I give to which component. and which component will have which callback function as parameter. like you suggested I have made shiftsPart an individual component and given it {setList} a parameter and then use {setList} as prop. again I am not clear about display component.

File: react/react1/week3/my-app/src/TodoList.js (hunk 3 above)
Comment (faheem102) [reply react/react1/week3/my-app/src/TodoList.js:160]: same problem here. not sure about parameter props. then which state should I use use inside this component. when I use inputdescription state and inputdeadline. it says it is undefine.

**Comment by faheem102** [comment]
hi bjørn i have tried to implement changes that you have suggested. problem now I have is my list is not searching. so the searchemploye function is not working inside ShiftPart component. also when I copy display function and paste it in the relative return it given me error and does not display the list. please look into the latest commit with the code.

**Comment by faheem102** [comment]
hi bjørn please have a look on my "implement suggested changes" commit. where I have successfully made individual components as you suggested. please let me know if I can merge the PR

**Comment by BjornFridal** [comment]
It looks good Faheem! You have broken the code into components and the search is working as expected 👌 Feel free to merge it.

**Pull request merged**